File: algorithms/triangle/triangle.cpp
// Source : https://oj.leetcode.com/problems/triangle/
// Author : Hao Chen
// Date   : 2014-06-18

/********************************************************************************** 
* 
* Given a triangle, find the minimum path sum from top to bottom. 
* Each step you may move to adjacent numbers on the row below.
* 
* For example, given the following triangle
* 
* [
*      [2],
*     [3,4],
*    [6,5,7],
*   [4,1,8,3]
* ]
* 
* The minimum path sum from top to bottom is 11 (i.e., 2 + 3 + 5 + 1 = 11).
* 
* Note:
* Bonus point if you are able to do this using only O(n) extra space, where n is the total number of rows in the triangle.
* 
*               
**********************************************************************************/

#include <iostream>
#include <vector>
using namespace std;

class Solution {
    
public:
    int minimumTotal(vector<vector<int> > &triangle) {
        vector< vector<int> > v;
        
        for (int i=0; i<triangle.size(); i++){
            
            if(i==0){
                v.push_back(triangle[i]);
                continue;
            }
            
            vector<int> tmp;
            
            
            for(int j=0; j<triangle[i].size(); j++){
                int x, y, z;
                x = y = z = 0x7fff;
                if ( (j-1) >= 0){
                    x = v[i-1][j-1];
                }
                if (j<v[i-1].size()) {
                    y = v[i-1][j];
                }
                /* won't take the previous adjacent number */
                //if ( (j+1)<v[i-1].size()) {
                //    z = v[i-1][j+1];
                //}
                tmp.push_back( min(x,y,z) + triangle[i][j] );
            }
            
            v.push_back(tmp);
            
        }
        int min=0x7fff;
        if (v.size() > 0){
            vector<int> &vb = v[v.size()-1];
            for(int i=0; i<vb.size(); i++){
                if (vb[i] < min ){
                    min = vb[i];
                }
            }
        }
           
        return min;
    }
private:
    inline int min(int x, int y, int z){
        int n = x<y?x:y;
        return (n<z?n:z);
    }
};


int main()
{
    vector< vector<int> > v;
    vector<int> i;
    i.push_back(-1);
    v.push_back(i);
    
    i.clear();
    i.push_back(2);
    i.push_back(3);
    v.push_back(i);

    i.clear();
    i.push_back(1);
    i.push_back(-1);
    i.push_back(-3);
    v.push_back(i);

    Solution s;
    cout << s.minimumTotal(v) << endl;;
    
    v.clear();
    i.clear();
    i.push_back(-1);
    v.push_back(i);
    
    i.clear();
    i.push_back(3);
    i.push_back(2);
    v.push_back(i);

    i.clear();
    i.push_back(-3);
    i.push_back(1);
    i.push_back(-1);
    v.push_back(i);
    cout << s.minimumTotal(v) << endl;;

    return 0;
}

File: algorithms-java/src/dynamicProgramming/triangle/triangle.java
// Source : https://oj.leetcode.com/problems/triangle/
// Inspired by : http://www.jiuzhang.com/solutions/triangle/
// Author : Lei Cao
// Date   : 2015-10-12

/**********************************************************************************
 *
 * Given a triangle, find the minimum path sum from top to bottom.
 * Each step you may move to adjacent numbers on the row below.
 *
 * For example, given the following triangle
 *
 * [
 *      [2],
 *     [3,4],
 *    [6,5,7],
 *   [4,1,8,3]
 * ]
 *
 * The minimum path sum from top to bottom is 11 (i.e., 2 + 3 + 5 + 1 = 11).
 *
 * Note:
 * Bonus point if you are able to do this using only O(n) extra space, where n is the total number of rows in the triangle.
 *
 *
 **********************************************************************************/

package dynamicProgramming.triangle;

import java.util.List;

/**
 * Created by leicao on 12/10/15.
 */
public class triangle {
    /**
     * @param triangle: a list of lists of integers.
     * @return: An integer, minimum path sum.
     */
    public int minimumTotal1(List<List<Integer>> triangle) {
        if (triangle == null) {
            return 0;
        }
        int rows = triangle.size();
        int[][] matrix = new int[rows][rows];
        matrix[0][0] = triangle.get(0).get(0);
        for (int i = 1; i < rows; i++) {
            matrix[i][0] = triangle.get(i).get(0) + matrix[i-1][0];
        }

        for (int i = 1; i < rows; i++) {
            matrix[i][i] = triangle.get(i).get(i) + matrix[i - 1][i - 1];
        }

        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < i; j++) {
                matrix[i][j] = Math.min(matrix[i-1][j-1], matrix[i-1][j]) + triangle.get(i).get(j);
            }
        }
        int minimum = matrix[rows - 1][0];
        for (int i = 1; i < rows; i++) {
            if (matrix[rows - 1][i] < minimum) {
                minimum = matrix[rows - 1][i];
            }
        }
        return minimum;
    }

    public int minimumTotal2(int[][] triangle) {
        if (triangle == null) {
            return 0;
        }
        int rows = triangle.length;
        int[][] matrix = new int[rows][rows];
        matrix[0][0] = triangle[0][0];
        for (int i = 1; i < rows; i++) {
            matrix[i][0] = triangle[i][0] + matrix[i-1][0];
        }

        for (int i = 1; i < rows; i++) {
            matrix[i][i] = triangle[i][i] + matrix[i - 1][i - 1];
        }

        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < i; j++) {
                matrix[i][j] = Math.min(matrix[i-1][j-1], matrix[i-1][j]) + triangle[i][j];
            }
        }
        int minimum = matrix[rows - 1][0];
        for (int i = 1; i < rows; i++) {
            if (matrix[rows - 1][i] < minimum) {
                minimum = matrix[rows - 1][i];
            }
        }
        return minimum;
    }
}
